### PR TITLE
add check plugin

### PR DESCRIPTION
**Pull request by rphillips**

### CODE REVIEW

File: check/check_plugin.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+	// Start and close stdin
+	cmd.Start()
+	r.Close()
+
+	stdoutReadDone := make(chan struct{})
+	go func() {
+		defer close(stdoutReadDone)
+		scanner := bufio.NewScanner(stdout)
+		statusRegex, _ := regexp.Compile("^status\\s+(err|warn|ok)\\s+(.*)")
```
Comment (itzg) [inline check/check_plugin.go]: You can also declare these as MustCompile and then put them at the top as file var's.

File: check/check_plugin.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+				metricName := matches[1]
+				metricType := strings.ToLower(matches[2])
+				metricValue := matches[3]
+				var pollerType int
+				switch metricType {
+				case "string":
+					pollerType = metric.MetricString
+				case "double":
+					pollerType = metric.MetricFloat
+				case "gauge":
```
Comment (itzg) [inline check/check_plugin.go]: I had discovered you can also comma-list these instead of fallthrough

File: check/check_plugin.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+				default:
+					pollerType = metric.MetricString
+				}
+				cr.AddMetric(metric.NewMetric(metricName, "", pollerType, metricValue, metricType))
+			}
+		}
+	}()
+
+	// Wait for commmand to finish
+	cmd.Wait()
```
Comment (itzg) [inline check/check_plugin.go]: I wonder what strategies they have for time bounding and force killing the process...

File: check/check_plugin.go (hunk 1 above)
Comment (rphillips) [inline check/check_plugin.go]: I changed them to MustCompile's. Great idea... I don't think these structures are thread safe though.

File: check/check_plugin.go (hunk 2 above)
Comment (rphillips) [inline check/check_plugin.go]: oh awesome! fixed.

File: check/check_plugin.go (hunk 3 above)
Comment (rphillips) [inline check/check_plugin.go]: context's FTW... It gets setup at the top.

File: check/check_plugin.go (hunk 1 above)
Comment (itzg) [inline check/check_plugin.go]: It was buried in the docs, but finally found it :)

> A Regexp is safe for concurrent use by multiple goroutines.

 -https://diigo.com/09mpae

File: check/check_tcp_test.go [hunk 4]
```diff
@@ -157,6 +157,10 @@ func TestTCPRunSuccess(t *testing.T) {
 }
 
 func TestTCPRunFailureClosedPort(t *testing.T) {
+	// TODO: This behavior has changed in 1.8.x versions of golang
```
Comment (itzg) [inline check/check_tcp_test.go]: Yeah, this always was a delicate one, so totally fair to skip for now

File: check/check_plugin.go (hunk 1 above)
Comment (rphillips) [inline check/check_plugin.go]: sweet! fixed in 46fe583

**Comment by rphillips** [comment]
Not sure if we will use this check yet, but it is one of the more difficult checks within the agent

File: check/check_plugin.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	log "github.com/Sirupsen/logrus"
+	protocol "github.com/racker/rackspace-monitoring-poller/protocol/check"
+)
+
+const (
+	ErrorPluginExit = "Plugin exited with non-zero status code"
+)
+
+var (
+	statusRegex = regexp.MustCompile("^status\\s+(err|warn|ok)\\s+(.*)")
```
Comment (itzg) [inline check/check_plugin.go]: Much nicer declaring up here, right? :)

File: check/check_plugin_test.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/racker/rackspace-monitoring-poller/check"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAgentPlugin_RunSuccess(t *testing.T) {
+	// Create Check
+	checkData := fmt.Sprintf(`{
```
Comment (itzg) [inline check/check_plugin_test.go]: Only if you want to, it'd save some lines of code to create a helper utility for creating these checkData blobs.

File: check/check_plugin.go (hunk 5 above)
Comment (rphillips) [inline check/check_plugin.go]: oh yeah. that is a good call

File: check/check_plugin_test.go (hunk 6 above)
Comment (rphillips) [inline check/check_plugin_test.go]: sometimes I want to tweak them.. especially the timeouts. I think it's fine for now.

**Pull request merged**